src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2364,1537 @@ IsForeignTable(Oid relationId)
+}
+
+
+/*
+ * GetNextBackgroundTaskTaskId reads and increments the SQL sequence associated with the
+ * background job tasks' task_id. After incrementing the counter it returns the counter
+ * back to the caller.
+ *
+ * The return value is typically used to insert new tasks into the catalog.
+ */
+int64
+GetNextBackgroundTaskTaskId(void)
+{
+	return DatumGetInt64(nextval_internal(DistBackgroundTasksTaskIdSequenceId(), false));
+}
+
+
+/*
+ * CreateBackgroundJob is a helper function to insert a new Background Job into Citus'
+ * catalog. After inserting the new job's metadataa into the catalog it returns the job_id
+ * assigned to the new job. This is typically used to associate new tasks with the newly
+ * created job.
+ */
+int64
+CreateBackgroundJob(const char *jobType, const char *description)
+{
+	Relation pgDistBackgroundJobs =
+		table_open(DistBackgroundJobsRelationId(), RowExclusiveLock);
+
+	/* insert new job */
+	Datum values[Natts_pg_dist_background_jobs] = { 0 };
+	bool nulls[Natts_pg_dist_background_jobs] = { 0 };
+
+	memset(nulls, true, sizeof(nulls));
+
+	int64 jobId = GetNextBackgroundJobsJobId();
+
+	values[Anum_pg_dist_background_jobs_job_id - 1] = Int64GetDatum(jobId);
+	nulls[Anum_pg_dist_background_jobs_job_id - 1] = false;
+
+	values[Anum_pg_dist_background_jobs_state - 1] = CitusJobStatusScheduledId();
+	nulls[Anum_pg_dist_background_jobs_state - 1] = false;
+
+	if (jobType)
+	{
+		NameData jobTypeName = { 0 };
+		namestrcpy(&jobTypeName, jobType);
+		values[Anum_pg_dist_background_jobs_job_type - 1] = NameGetDatum(&jobTypeName);
+		nulls[Anum_pg_dist_background_jobs_job_type - 1] = false;
+	}
+
+	if (description)
+	{
+		values[Anum_pg_dist_background_jobs_description - 1] =
+			CStringGetTextDatum(description);
+		nulls[Anum_pg_dist_background_jobs_description - 1] = false;
+	}
+
+	HeapTuple newTuple = heap_form_tuple(RelationGetDescr(pgDistBackgroundJobs),
+										 values, nulls);
+	CatalogTupleInsert(pgDistBackgroundJobs, newTuple);
+
+	table_close(pgDistBackgroundJobs, NoLock);
+
+	return jobId;
+}
+
+
+/*
+ * ScheduleBackgroundTask creates a new background task to be executed in the background.
+ *
+ * The new task is associated with an existing job based on it's id.
+ *
+ * Optionally the new task can depend on separate tasks associated with the same job. When
+ * a new task is created with dependencies on previous tasks we assume this task is
+ * blocked on its depending tasks.
+ */
+BackgroundTask *
+ScheduleBackgroundTask(int64 jobId, Oid owner, char *command, int dependingTaskCount,
+					   int64 dependingTaskIds[])
+{
+	BackgroundTask *task = NULL;
+
+	Relation pgDistBackgroundTasks =
+		table_open(DistBackgroundTasksRelationId(), RowExclusiveLock);
+	Relation pgDistbackgroundTasksDepend = NULL;
+	if (dependingTaskCount > 0)
+	{
+		pgDistbackgroundTasksDepend =
+			table_open(DistBackgroundTasksDependRelationId(), RowExclusiveLock);
+	}
+
+	/* 1. TODO verify depending jobs exist and lock them */
```

Comment:
There's a TODO here. Should that happen before merging this PR?

Reply:
Some basic sanity checks there would be good I think, especially since it is required to guarantee the FK's are not violated. Given that Catalog inserts/updates don't run FK checks.

Reply:
Locking will be hard, as it is not apparent how to lock a row in a catalog table. Postgres mostly relies on other locks for the objects to be obtained when updating the catalog. Citus mostly works around it as it seems.

The only way would be to exclusively lock the tasks table, which we do now during cancel and updates. Probably we can extend that here and simply serialize any job related operations. I will atleast do some basic sanity checking now and we can discuss the serialized access during tomorrows sync.

Reply:
This is fully resolved now by taking `ExclusiveLock` on the catalog tables while adding records and checking constraints.